keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1338 @@
+
+### Risks and Mitigations
+
+Users may configure both parameters in StorageClass and parameters in VolumeAttributesClass differently. The mitigation here is to provide a guideline of solving conflicts. If the MODIFY\_VOLUME capability is present, VolumeAttributesClass should be honored.
+
+## Design Details
+
+As part of this proposal, we are proposing:
+
+
+### 1. Add Create/Delete VolumeAttributesClass support in Kubernetes including vac\_admission\_controller and vac\_protection\_controller for deletion protection. 
+
+
+#### VolumeAttributesClass Deletion Protection
+
+While a VolumeAttributesClass is referenced by any PVC, we will prevent the object from being deleted by adding a finalizer([reference](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go)). 
+
+The VACObjectInUseProtection admission controller sets the finalizer on all VolumeAttributesClasses. VACProtectionController removes the finalizer when it's not referenced. This prevents users from deleting a VolumeAttributesClass that's used by a PVC.
+
+There are a few conditions that will trigger add/remove pvc finalizers in the VolumeAttributesClass:
+
+
+
+1. PVC created with a VolumeAttributesClass
+
+    The **VACObjectInUseProtection admission controller**:
+    * Check if this VolumeAttributesClass already has a protection finalizer
+    * Add the finalizer to the VolumeAttributesClass if there is none 
+2. PVC created with a VolumeAttributesClass being deleted
+    The **VACObjectInUseProtection admission controller**:
+    * Check VolumeAttributesClass is being deleted and PVC creation failed
+3. PVC updated to a different VolumeAttributesClass
+    * The **VACProtectionController** will remove PVC finalizer in previous VolumeAttributesClass if after listing all the PVCs and confirmed that this PVC is the last one that is consuming the previous VolumeAttributesClass
+
+    The **VACObjectInUseProtection admission controller**:
+    * Check if the new VolumeAttributesClass already has a protection finalizer
+    * Add the finalizer to the new VolumeAttributesClass if there is none 
+4. PVC updated to a different VolumeAttributesClass that is being deleted
+    The **VACObjectInUseProtection admission controller**:
+    * Check VolumeAttributesClass is being deleted and PVC update failed
+5. PVC has a VolumeAttributesClass and this PVC is deleted
+    * The **VACProtectionController** will remove finalizer in the VolumeAttributesClass only if after listing all the PVCs and confirm that this PVC is the last one that is consuming the VolumeAttributesClass
+6. Delete a VolumeAttributesClass while there is **kubernetes.io/vac-protection** finalizer associated with this VolumeAttributesClass
+    * Deletion failed
```

Comment:
Once the final reference is gone, the finalizer will be removed and then the object will be deleted. There is no need to delete again because the object already has a deletionTimestamp on it.